api/jobs.go

```diff
@@ -24,84 +29,125 @@ var (
 )
 
-// CreateJobHandler returns a function that generates a new Job resource containing default values in its fields.
-func (api *API) CreateJobHandler(ctx context.Context) http.HandlerFunc {
-	return func(w http.ResponseWriter, req *http.Request) {
-		ctx := req.Context()
+// CreateJobHandler generates a new Job resource and a new ElasticSearch index associated with it	.
+func (api *API) CreateJobHandler(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	id := NewID()
 
-		id := NewID()
+	log.Info(ctx, "creating new job resource in the data store")
+	newJob, err := api.dataStore.CreateJob(ctx, id)
+	if err != nil {
+		log.Error(ctx, "creating and storing job failed", err)
+		if err == mongo.ErrExistingJobInProgress {
+			http.Error(w, "existing reindex job in progress", http.StatusConflict)
+		} else {
+			http.Error(w, serverErrorMessage, http.StatusInternalServerError)
+		}
+		return
+	}
 
-		newJob, err := api.dataStore.CreateJob(ctx, id)
-		if err != nil {
-			log.Error(ctx, "creating and storing job failed", err)
-			if err == mongo.ErrExistingJobInProgress {
-				http.Error(w, "existing reindex job in progress", http.StatusConflict)
-			} else {
+	log.Info(ctx, "creating new index in ElasticSearch via the Search API")
+	serviceAuthToken := "Bearer " + api.cfg.ServiceAuthToken
+	searchAPISearchURL := api.cfg.SearchAPIURL + "/search"
+	reindexResponse, err := api.reindex.CreateIndex(ctx, serviceAuthToken, searchAPISearchURL, api.httpClient)
+	if err != nil {
+		log.Error(ctx, "error occurred when connecting to Search API", err)
+		if newJob != (models.Job{}) {
+			newJob.State = jobStateFailed
+			log.Info(ctx, "updating job state to failed", log.Data{"job id": newJob.ID})
+			setStateErr := api.dataStore.UpdateJobState(jobStateFailed, newJob.ID)
+			if setStateErr != nil {
+				log.Error(ctx, "setting state to failed has failed", setStateErr)
 				http.Error(w, serverErrorMessage, http.StatusInternalServerError)
+				return
 			}
-			return
 		}
-
-		w.Header().Set("Content-Type", "application/json")
-		jsonResponse, err := json.Marshal(newJob)
-		if err != nil {
-			log.Error(ctx, "marshalling response failed", err)
-			http.Error(w, serverErrorMessage, http.StatusInternalServerError)
-			return
+	} else if reindexResponse.StatusCode != 201 {
+		log.Error(ctx, "error occurred in post search http request", err)
+		if newJob != (models.Job{}) {
+			newJob.State = jobStateFailed
+			log.Info(ctx, "updating job state to failed", log.Data{"job id": newJob.ID})
+			setStateErr := api.dataStore.UpdateJobState(jobStateFailed, newJob.ID)
+			if setStateErr != nil {
+				log.Error(ctx, "setting state to failed has failed", setStateErr)
+				http.Error(w, serverErrorMessage, http.StatusInternalServerError)
+				return
+			}
 		}
-
-		w.WriteHeader(http.StatusCreated)
-		_, err = w.Write(jsonResponse)
+	} else {
+		newJob, err = api.updateSearchIndexName(ctx, reindexResponse, err, newJob, id)
```

Comment:
`err` passed in here will always be nil won't it? Can we just pass `nil` to clarify this?

Reply:
maybe next time

Reply:
now removed from function so no need to pass in

Reply:
looks good